File: test/parse_test.py
import os
import unittest
import sxpb

class TestSxpbFiles(unittest.TestCase):
    def test_validate_sxpb_files(self):
        dirs_file = os.path.join(os.path.dirname(__file__), 'dirs.sxpb')
        config = sxpb.load(dirs_file)
        dirs_to_scan = config.get('dirs', [])

        sxpb_files = []
        # If we are in root, dirs are relative to root.
        # But we should be careful about where the test is run from.
        # Assuming run from root as per `pdm test`.

        for d in dirs_to_scan:
            if not os.path.exists(d):
                continue
            for root, _, files in os.walk(d):
                for file in files:
                    if file.endswith('.sxpb'):
                        sxpb_files.append(os.path.join(root, file))

        self.assertTrue(len(sxpb_files) > 0, "No .sxpb files found")

        for sxpb_file in sxpb_files:
            with self.subTest(sxpb_file=sxpb_file):
                try:
                    # 1. Read the file via the library, using precise=True
                    obj1 = sxpb.load(sxpb_file, precise=True)

                    # 2. Write the sxpb to a string
                    s1 = sxpb.dumps(obj1)

                    # 3. Parse sxpb from the string, using precise=True
                    obj2 = sxpb.loads(s1, precise=True)

                    # 4. Write the sxpb to another string
                    s2 = sxpb.dumps(obj2)

                    # 5. Compare the 2 written strings
                    self.assertEqual(s1, s2, f"Idempotency check failed for {sxpb_file}")

                except Exception as e:
                    self.fail(f"Failed to process {sxpb_file}: {e}")

if __name__ == '__main__':
    unittest.main()

File: test/parse_test.js
import fs from "fs";
import path from "path";
import SxPB from "@sxproto/sxpb";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const dirsFile = path.join(__dirname, "dirs.sxpb");
const dirsToScan = SxPB.parse(fs.readFileSync(dirsFile, "utf8"));

let sxpbFiles = [];

function walk(dir) {
  if (!fs.existsSync(dir)) return;
  const list = fs.readdirSync(dir);
  list.forEach(function (file) {
    file = path.join(dir, file);
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      walk(file);
    } else {
      if (file.endsWith(".sxpb")) {
        sxpbFiles.push(file);
      }
    }
  });
}

const projectRoot = path.join(__dirname, "..");

dirsToScan.forEach((dir) => {
  const fullPath = path.join(projectRoot, dir);
  walk(fullPath);
});

if (sxpbFiles.length === 0) {
  console.error("No .sxpb files found");
  process.exit(1);
}

let failed = false;

sxpbFiles.forEach((file) => {
  try {
    // 1. Read the file via the library, using precise=true
    const content = fs.readFileSync(file, "utf8");
    const obj1 = SxPB.parse(content, { precise: true });

    // 2. Write the sxpb to a string
    const s1 = SxPB.stringify(obj1);

    // 3. Parse sxpb from the string, using precise=true
    const obj2 = SxPB.parse(s1, { precise: true });

    // 4. Write the sxpb to another string
    const s2 = SxPB.stringify(obj2);

    // 5. Compare the 2 written strings
    if (s1 !== s2) {
      throw new Error(
        `Idempotency check failed. Expected length ${s1.length}, got ${s2.length}`,
      );
    }

    console.log(`Validated ${file}`);
  } catch (error) {
    console.error(`Validation failed for ${file}:`, error);
    failed = true;
  }
});

if (failed) {
  process.exit(1);
}
